File: DnD/static/js/build-calculator.js
// Build Calculator JavaScript

let builds = [];
let dprChart = null;

// Weapon damage data
const weapons = {
    'longbow': { dice: 8, count: 1, name: 'Longbow' },
    'shortbow': { dice: 6, count: 1, name: 'Shortbow' },
    'hand_crossbow': { dice: 6, count: 1, name: 'Hand Crossbow' },
    'light_crossbow': { dice: 8, count: 1, name: 'Light Crossbow' },
    'heavy_crossbow': { dice: 10, count: 1, name: 'Heavy Crossbow' }
};

// Calculate ability modifier
function calculateModifier(abilityScore) {
    return Math.floor((abilityScore - 10) / 2);
}

// Calculate proficiency bonus
function calculateProficiencyBonus(level) {
    return 2 + Math.floor((level - 1) / 4);
}

// Calculate to-hit bonus
function calculateToHit(level, abilityMod, magicBonus, sharpshooter) {
    const profBonus = calculateProficiencyBonus(level);
    const penalty = sharpshooter ? -5 : 0;
    return abilityMod + profBonus + magicBonus + penalty;
}

// Calculate hit chance
function calculateHitChance(toHit, targetAC) {
    const neededRoll = targetAC - toHit;
    
    if (neededRoll <= 1) return 0.95; // Always hit except nat 1
    if (neededRoll >= 20) return 0.05; // Only hit on nat 20
    
    return (21 - neededRoll) / 20;
}

// Calculate DPR
function calculateDPR(buildConfig) {
    const level = parseInt(buildConfig.level);
    const abilityScore = parseInt(buildConfig.abilityScore);
    const abilityMod = calculateModifier(abilityScore);
    const weapon = weapons[buildConfig.weapon];
    const magicBonus = parseInt(buildConfig.magicBonus);
    const sharpshooter = buildConfig.sharpshooter;
    const crossbowExpert = buildConfig.crossbowExpert;
    const attacksPerRound = parseInt(buildConfig.attacksPerRound);
    const targetAC = parseInt(buildConfig.targetAC);
    
    // Calculate to-hit
    const toHit = calculateToHit(level, abilityMod, magicBonus, sharpshooter);
    
    // Calculate damage per hit
    const weaponDamageAvg = weapon.count * ((weapon.dice + 1) / 2);
    const sharpshooterDamage = sharpshooter ? 10 : 0;
    const damagePerHit = weaponDamageAvg + abilityMod + magicBonus + sharpshooterDamage;
    
    // Calculate hit chance
    const hitChance = calculateHitChance(toHit, targetAC);
    
    // Critical hits
    const critDamage = weaponDamageAvg;
    const critChance = 0.05;
    
    // Expected damage per attack
    const expectedPerAttack = (hitChance * damagePerHit) + (critChance * critDamage);
    
    // Total attacks
    let totalAttacks = attacksPerRound;
    if (crossbowExpert && buildConfig.weapon === 'hand_crossbow') {
        totalAttacks += 1; // Bonus action attack
    }
    
    const dpr = expectedPerAttack * totalAttacks;
    
    return {
        dpr: Math.round(dpr * 100) / 100,
        toHit: toHit,
        hitChance: Math.round(hitChance * 1000) / 10,
        damagePerHit: Math.round(damagePerHit * 100) / 100,
        attacksPerRound: totalAttacks,
        expectedPerAttack: Math.round(expectedPerAttack * 100) / 100
    };
}

// Add build to comparison
function addBuild() {
    const buildConfig = {
        name: document.getElementById('buildName').value,
        level: document.getElementById('level').value,
        class: document.getElementById('classSelect').value,
        subclass: document.getElementById('subclassSelect').value,
        abilityScore: document.getElementById('abilityScore').value,
        weapon: document.getElementById('weaponSelect').value,
        magicBonus: document.getElementById('magicBonus').value,
        sharpshooter: document.getElementById('sharpshooter').checked,
        crossbowExpert: document.getElementById('crossbowExpert').checked,
        attacksPerRound: document.getElementById('attacksPerRound').value,
        targetAC: document.getElementById('targetAC').value
    };
    
    const results = calculateDPR(buildConfig);
    
    builds.push({
        config: buildConfig,
        results: results
    });
    
    updateComparison();
    updateChart();
}

// Update comparison display
function updateComparison() {
    const container = document.getElementById('comparisonResults');
    
    if (builds.length === 0) {
        container.innerHTML = '<p class="text-muted text-center">Add builds to compare them side by side</p>';
        return;
    }
    
    // Sort by DPR
    builds.sort((a, b) => b.results.dpr - a.results.dpr);
    
    let html = '<div class="table-responsive"><table class="table table-striped table-hover">';
    html += '<thead><tr>';
    html += '<th>Rank</th><th>Build Name</th><th>DPR</th><th>To Hit</th><th>Hit %</th>';
    html += '<th>Damage/Hit</th><th>Attacks</th><th>Details</th><th>Action</th>';
    html += '</tr></thead><tbody>';
    
    builds.forEach((build, index) => {
        const config = build.config;
        const results = build.results;
        
        const feats = [];
        if (config.sharpshooter) feats.push('Sharpshooter');
        if (config.crossbowExpert) feats.push('Crossbow Expert');
        
        html += '<tr>';
        html += `<td><strong>#${index + 1}</strong></td>`;
        html += `<td><strong>${config.name}</strong><br><small class="text-muted">Lvl ${config.level} ${config.subclass}</small></td>`;
        html += `<td><span class="badge bg-success fs-6">${results.dpr}</span></td>`;
        html += `<td>+${results.toHit}</td>`;
        html += `<td>${results.hitChance}%</td>`;
        html += `<td>${results.damagePerHit}</td>`;
        html += `<td>${results.attacksPerRound}</td>`;
        html += `<td><small>${weapons[config.weapon].name}`;
        if (config.magicBonus > 0) html += ` +${config.magicBonus}`;
        if (feats.length > 0) html += `<br>${feats.join(', ')}`;
        html += `</small></td>`;
        html += `<td><button class="btn btn-sm btn-danger" onclick="removeBuild(${index})"><i class="bi bi-trash"></i></button></td>`;
        html += '</tr>';
    });
    
    html += '</tbody></table></div>';
    
    // Add insights
    if (builds.length >= 2) {
        const best = builds[0];
        const dprDiff = best.results.dpr - builds[1].results.dpr;
        
        html += '<div class="alert alert-info mt-3">';
        html += `<strong><i class="bi bi-lightbulb"></i> Analysis:</strong> `;
        html += `<strong>${best.config.name}</strong> deals ${dprDiff.toFixed(2)} more DPR than the next best build. `;
        
        if (best.results.hitChance < 60) {
            html += 'However, the hit chance is below 60% - consider dropping Sharpshooter for more consistent damage. ';
        }
        
        html += '</div>';
    }
    
    container.innerHTML = html;
}

// Update DPR chart
function updateChart() {
    const ctx = document.getElementById('dprChart');
    
    if (dprChart) {
        dprChart.destroy();
    }
    
    const labels = builds.map(b => b.config.name);
    const dprData = builds.map(b => b.results.dpr);
    const hitChanceData = builds.map(b => b.results.hitChance);
    
    dprChart = new Chart(ctx, {
        type: 'bar',
        data: {
            labels: labels,
            datasets: [
                {
                    label: 'DPR',
                    data: dprData,
                    backgroundColor: 'rgba(40, 167, 69, 0.7)',
                    borderColor: 'rgba(40, 167, 69, 1)',
                    borderWidth: 2,
                    yAxisID: 'y'
                },
                {
                    label: 'Hit Chance %',
                    data: hitChanceData,
                    backgroundColor: 'rgba(0, 123, 255, 0.7)',
                    borderColor: 'rgba(0, 123, 255, 1)',
                    borderWidth: 2,
                    yAxisID: 'y1'
                }
            ]
        },
        options: {
            responsive: true,
            interaction: {
                mode: 'index',
                intersect: false
            },
            scales: {
                y: {
                    type: 'linear',
                    display: true,
                    position: 'left',
                    title: {
                        display: true,
                        text: 'DPR'
                    }
                },
                y1: {
                    type: 'linear',
                    display: true,
                    position: 'right',
                    title: {
                        display: true,
                        text: 'Hit Chance %'
                    },
                    grid: {
                        drawOnChartArea: false
                    }
                }
            }
        }
    });
}

// Remove build
function removeBuild(index) {
    builds.splice(index, 1);
    updateComparison();
    updateChart();
}

// Load template builds
function loadTemplate(templateName) {
    const templates = {
        'valorSharpshooter': {
            name: 'Valor Sharpshooter',
            level: 8,
            class: 'bard',
            subclass: 'valor',
            abilityScore: 18,
            weapon: 'longbow',
            magicBonus: 1,
            sharpshooter: true,
            crossbowExpert: false,
            attacksPerRound: 2,
            targetAC: 16
        },
        'valorCrossbowExpert': {
            name: 'Valor Crossbow Expert',
            level: 8,
            class: 'bard',
            subclass: 'valor',
            abilityScore: 18,
            weapon: 'hand_crossbow',
            magicBonus: 1,
            sharpshooter: false,
            crossbowExpert: true,
            attacksPerRound: 2,
            targetAC: 16
        },
        'loreBard': {
            name: 'Lore Bard (Spell Focus)',
            level: 8,
            class: 'bard',
            subclass: 'lore',
            abilityScore: 20,
            weapon: 'light_crossbow',
            magicBonus: 0,
            sharpshooter: false,
            crossbowExpert: false,
            attacksPerRound: 1,
            targetAC: 16
        },
        'swordsBard': {
            name: 'Swords Bard',
            level: 8,
            class: 'bard',
            subclass: 'swords',
            abilityScore: 18,
            weapon: 'shortbow',
            magicBonus: 0,
            sharpshooter: false,
            crossbowExpert: false,
            attacksPerRound: 2,
            targetAC: 16
        }
    };
    
    const template = templates[templateName];
    if (!template) return;
    
    document.getElementById('buildName').value = template.name;
    document.getElementById('level').value = template.level;
    document.getElementById('classSelect').value = template.class;
    document.getElementById('subclassSelect').value = template.subclass;
    document.getElementById('abilityScore').value = template.abilityScore;
    document.getElementById('weaponSelect').value = template.weapon;
    document.getElementById('magicBonus').value = template.magicBonus;
    document.getElementById('sharpshooter').checked = template.sharpshooter;
    document.getElementById('crossbowExpert').checked = template.crossbowExpert;
    document.getElementById('attacksPerRound').value = template.attacksPerRound;
    document.getElementById('targetAC').value = template.targetAC;
}

File: DnD/utils/build_calculator.py
"""
Build Calculator and DPR Calculator
Calculates damage per round for different builds
"""


def calculate_modifier(ability_score):
    """Calculate ability modifier"""
    return (ability_score - 10) // 2


def calculate_to_hit(level, ability_mod, proficiency=True, magic_bonus=0, feat_penalty=0):
    """Calculate to-hit bonus"""
    prof_bonus = 2 + ((level - 1) // 4) if proficiency else 0
    return ability_mod + prof_bonus + magic_bonus + feat_penalty


def calculate_damage(
    weapon_damage_dice,
    weapon_damage_count,
    ability_mod,
    magic_bonus=0,
    extra_damage=0,
    extra_damage_dice=0,
    extra_damage_count=0
):
    """Calculate average damage per hit"""
    # Average of a die is (max + 1) / 2
    weapon_avg = weapon_damage_count * ((weapon_damage_dice + 1) / 2)
    extra_avg = extra_damage_count * ((extra_damage_dice + 1) / 2) if extra_damage_dice > 0 else 0
    
    total_damage = weapon_avg + ability_mod + magic_bonus + extra_damage + extra_avg
    return total_damage


def calculate_hit_chance(to_hit_bonus, target_ac=15):
    """Calculate hit probability"""
    needed_roll = target_ac - to_hit_bonus
    
    if needed_roll <= 1:
        return 0.95  # Always hit except on nat 1
    elif needed_roll >= 20:
        return 0.05  # Only hit on nat 20
    else:
        return (21 - needed_roll) / 20


def calculate_dpr(
    level,
    ability_score,
    weapon_damage_dice,
    weapon_damage_count=1,
    attacks_per_round=1,
    bonus_action_attack=False,
    magic_weapon_bonus=0,
    sharpshooter=False,
    extra_damage_per_hit=0,
    extra_damage_dice=0,
    extra_damage_count=0,
    target_ac=15,
    advantage=False,
    disadvantage=False
):
    """
    Calculate Damage Per Round
    
    Args:
        level: Character level
        ability_score: Primary attacking ability (DEX or STR)
        weapon_damage_dice: Die size (6 for d6, 8 for d8, etc.)
        weapon_damage_count: Number of dice
        attacks_per_round: Number of attacks from Extra Attack
        bonus_action_attack: Whether you get a bonus action attack
        magic_weapon_bonus: +1/+2/+3 weapon
        sharpshooter: Using Sharpshooter -5/+10
        extra_damage_per_hit: Flat damage bonus (like Sneak Attack, Divine Smite)
        extra_damage_dice: Extra damage die size
        extra_damage_count: Number of extra damage dice
        target_ac: Enemy armor class
        advantage: Rolling with advantage
        disadvantage: Rolling with disadvantage
    
    Returns:
        dict with DPR breakdown
    """
    ability_mod = calculate_modifier(ability_score)
    
    # Calculate to-hit
    feat_penalty = -5 if sharpshooter else 0
    to_hit = calculate_to_hit(level, ability_mod, True, magic_weapon_bonus, feat_penalty)
    
    # Calculate damage per hit
    sharpshooter_damage = 10 if sharpshooter else 0
    damage_per_hit = calculate_damage(
        weapon_damage_dice,
        weapon_damage_count,
        ability_mod,
        magic_weapon_bonus,
        sharpshooter_damage + extra_damage_per_hit,
        extra_damage_dice,
        extra_damage_count
    )
    
    # Calculate hit chance
    base_hit_chance = calculate_hit_chance(to_hit, target_ac)
    
    if advantage:
        hit_chance = 1 - ((1 - base_hit_chance) ** 2)
    elif disadvantage:
        hit_chance = base_hit_chance ** 2
    else:
        hit_chance = base_hit_chance
    
    # Critical hits (5% chance, double weapon dice only)
    crit_damage = weapon_damage_count * ((weapon_damage_dice + 1) / 2)
    if extra_damage_dice > 0:
        crit_damage += extra_damage_count * ((extra_damage_dice + 1) / 2)
    
    # Expected damage per attack
    expected_damage_per_attack = (hit_chance * damage_per_hit) + (0.05 * crit_damage)
    
    # Total attacks
    total_attacks = attacks_per_round
    if bonus_action_attack:
        total_attacks += 1
    
    total_dpr = expected_damage_per_attack * total_attacks
    
    return {
        'dpr': round(total_dpr, 2),
        'to_hit': to_hit,
        'hit_chance': round(hit_chance * 100, 1),
        'damage_per_hit': round(damage_per_hit, 2),
        'attacks_per_round': total_attacks,
        'expected_per_attack': round(expected_damage_per_attack, 2)
    }


def compare_builds(builds):
    """
    Compare multiple builds side by side
    
    Args:
        builds: List of build configurations
    
    Returns:
        Comparison dict
    """
    results = []
    
    for build in builds:
        dpr_result = calculate_dpr(**build['stats'])
        
        results.append({
            'name': build['name'],
            'description': build.get('description', ''),
            'dpr': dpr_result['dpr'],
            'to_hit': dpr_result['to_hit'],
            'hit_chance': dpr_result['hit_chance'],
            'utility': build.get('utility', []),
            'spell_slots': build.get('spell_slots', {}),
            'breakdown': dpr_result
        })
    
    # Sort by DPR
    results.sort(key=lambda x: x['dpr'], reverse=True)
    
    return results


# Example: Ranged Bard Builds
if __name__ == '__main__':
    # College of Valor Bard with Sharpshooter
    valor_sharpshooter = {
        'name': 'Valor Bard - Sharpshooter',
        'description': 'College of Valor, Longbow, Sharpshooter feat',
        'stats': {
            'level': 8,
            'ability_score': 18,  # 16 base + 2 ASI
            'weapon_damage_dice': 8,
            'weapon_damage_count': 1,
            'attacks_per_round': 2,  # Extra Attack from Valor
            'sharpshooter': True,
            'magic_weapon_bonus': 1,
            'target_ac': 16
        },
        'utility': ['Bardic Inspiration', 'Full spellcasting', 'Jack of All Trades'],
        'spell_slots': {1: 4, 2: 3, 3: 3, 4: 2}
    }
    
    # College of Lore Bard with Magic Initiate
    lore_versatile = {
        'name': 'Lore Bard - Spell Focus',
        'description': 'College of Lore, Light Crossbow, focus on spells',
        'stats': {
            'level': 8,
            'ability_score': 20,  # 16 + 2 ASI to max CHA
            'weapon_damage_dice': 8,
            'weapon_damage_count': 1,
            'attacks_per_round': 1,
            'target_ac': 16
        },
        'utility': ['Bardic Inspiration', 'Cutting Words', 'Full spellcasting', 'Extra Magical Secrets', 'Jack of All Trades'],
        'spell_slots': {1: 4, 2: 3, 3: 3, 4: 2}
    }
    
    # College of Valor with Crossbow Expert
    valor_crossbow = {
        'name': 'Valor Bard - Crossbow Expert',
        'description': 'College of Valor, Hand Crossbow, Crossbow Expert',
        'stats': {
            'level': 8,
            'ability_score': 18,
            'weapon_damage_dice': 6,
            'weapon_damage_count': 1,
            'attacks_per_round': 2,  # Extra Attack
            'bonus_action_attack': True,  # Crossbow Expert
            'magic_weapon_bonus': 1,
            'target_ac': 16
        },
        'utility': ['Bardic Inspiration', 'Full spellcasting', 'Jack of All Trades'],
        'spell_slots': {1: 4, 2: 3, 3: 3, 4: 2}
    }
    
    builds = [valor_sharpshooter, lore_versatile, valor_crossbow]
    comparison = compare_builds(builds)
    
    print("=== RANGED BARD BUILD COMPARISON ===\n")
    for i, build in enumerate(comparison, 1):
        print(f"{i}. {build['name']}")
        print(f"   DPR: {build['dpr']}")
        print(f"   To Hit: +{build['to_hit']} ({build['hit_chance']}% hit chance)")
        print(f"   Description: {build['description']}")
        print(f"   Utility: {', '.join(build['utility'])}")
        print(f"   Spell Slots: {build['spell_slots']}")
        print(f"   Breakdown: {build['breakdown']}")
        print()
